packages/core/src/emitter.ts

```diff
@@ -0,0 +1,311 @@
+	integrationUpdate: [WithIntrinsicProps<{ integration: GatewayIntegrationUpdateDispatchData }>];
+	interactionCreate: [WithIntrinsicProps<{ interaction: APIInteraction }>];
+	inviteCreate: [WithIntrinsicProps<{ invite: GatewayInviteCreateDispatchData }>];
+	inviteDelete: [WithIntrinsicProps<{ data: GatewayInviteDeleteDispatchData }>];
+	messageCreate: [WithIntrinsicProps<{ message: APIMessage }>];
+	messageDelete: [WithIntrinsicProps<{ data: GatewayMessageDeleteDispatchData }>];
+	messageDeleteBulk: [WithIntrinsicProps<{ data: GatewayMessageDeleteBulkDispatchData }>];
+	messageReactionAdd: [WithIntrinsicProps<{ data: GatewayMessageReactionAddDispatchData }>];
+	messageReactionRemove: [WithIntrinsicProps<{ data: GatewayMessageReactionRemoveDispatchData }>];
+	messageReactionRemoveAll: [WithIntrinsicProps<{ data: GatewayMessageReactionRemoveAllDispatchData }>];
+	messageReactionRemoveEmoji: [WithIntrinsicProps<{ data: GatewayMessageReactionRemoveEmojiDispatchData }>];
+	messageUpdate: [WithIntrinsicProps<{ message: GatewayMessageUpdateDispatchData }>];
+	presenceUpdate: [WithIntrinsicProps<{ data: GatewayPresenceUpdate }>];
+	ready: [WithIntrinsicProps<{ data: GatewayReadyDispatchData }>];
+	stageInstanceCreate: [WithIntrinsicProps<{ stage_instance: APIStageInstance }>];
+	stageInstanceDelete: [WithIntrinsicProps<{ stage_instance: APIStageInstance }>];
+	stageInstanceUpdate: [WithIntrinsicProps<{ stage_instance: APIStageInstance }>];
+	threadCreate: [WithIntrinsicProps<{ thread: APIThreadChannel }>];
+	threadDelete: [WithIntrinsicProps<{ thread: APIThreadChannel }>];
+	threadListSync: [WithIntrinsicProps<{ data: GatewayThreadListSync }>];
+	threadMemberUpdate: [WithIntrinsicProps<{ data: GatewayThreadMemberUpdateDispatchData }>];
+	threadMembersUpdate: [WithIntrinsicProps<{ data: GatewayThreadMembersUpdateDispatchData }>];
+	threadUpdate: [WithIntrinsicProps<{ thread: APIThreadChannel }>];
+	userUpdate: [WithIntrinsicProps<{ user: APIUser }>];
+	voiceServerUpdate: [WithIntrinsicProps<{ data: GatewayVoiceServerUpdateDispatchData }>];
+	voiceStateUpdate: [WithIntrinsicProps<{ state: GatewayVoiceState }>];
+	webhooksUpdate: [WithIntrinsicProps<{ data: GatewayWebhooksUpdateDispatchData }>];
+}
+
+export type ManagerShardEventsMap = {
+	[K in keyof MappedEvents]: MappedEvents[K];
+};
+
+export interface ClientOptions {
+	rest: REST;
+	ws: WebSocketManager;
+}
+
+export function createEmitter({ rest, ws }: ClientOptions) {
+	const eventAPI = new API(rest);
+	const emitter = new AsyncEventEmitter<ManagerShardEventsMap>();
+
+	ws.on(WebSocketShardEvents.Dispatch, ({ data: dispatch, shardId }) => {
+		function wrapIntrinsicProps<T>(obj: T): WithIntrinsicProps<T> {
+			return {
+				api: eventAPI,
+				shardId,
+				...obj,
+			};
+		}
+
+		emitter.emit('dispatch', wrapIntrinsicProps({ data: dispatch }));
+		switch (dispatch.t) {
+			case GatewayDispatchEvents.MessageCreate:
+				emitter.emit('messageCreate', wrapIntrinsicProps({ message: dispatch.d }));
```

Comment:
Just a question, any reason you don't emit the event based on what payload.t is, and instead rename it to a string (that has no enum either, something we've kinda been moving towards in djs)?